crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,707 @@
+use std::marker::PhantomData;
+
+use bevy_utils::{get_short_name, HashSet};
+
+use crate::{
+    component::{ComponentId, Components},
+    prelude::Bundle,
+    world::World,
+};
+
+/// A rule about which [`Component`](crate::component::Component)s can coexist on entities.
+///
+/// These rules must be true at all times for all entities in the [`World`].
+/// The generic [`Bundle`] type `B1` is always used in the `premise`,
+/// while `B2` is used in the `consequence`.
+/// If only a single generic is provided, these types are the same.
+///
+/// When added to the [`World`], archetype invariants behave like [`assert!`].
+/// Archetype invariants are checked each time [`Archetypes`](crate::archetype::Archetypes) is modified;
+/// this can occur on component addition, component removal, and entity spawning.
+///
+/// Archetypes are only modified when a novel archetype (set of components) is seen for the first time;
+/// swapping between existing archetypes will not trigger these checks.
+///
+/// Note that archetype invariants are not symmetric by default.
+/// For example, `ArchetypeInvariant::<B1, B2>::requires_one()` means that `B1` requires `B2`,
+/// but not that `B2` requires `B1`.
+/// In this case, an entity with just `B2` is completely valid, but an entity with just `B1` is not.
+/// If symmetry is desired, repeat the invariant with the order of the types switched.
+///
+/// When working with dynamic component types (for non-Rust components),
+/// use the untyped equivalents [`UntypedArchetypeInvariant`] and [`UntypedArchetypeStatement`] directly.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArchetypeInvariant<B1: Bundle, B2: Bundle = B1> {
```

Comment:
Something else I thought of and haven't really come to a decision on- is `ArchetypeInvariant` really the best name? I feel like `ArchetypeRequirement` would be an improvement at least for approachability. Right now it's just complete jargon, but if a newbie could at least figure out what `Archetype` means, they should be able to at least infer roughly what `ArchetypeRequirement` is. 
I'm not certain if this is definitely a positive change or if there might be a better name, but right now I'm slightly in favor of changing to `ArchetypeRequirement`(s).

Reply:
I'd prefer Rule or Assertion if we're trying to reduce technical jargon.

Current favorite is `ArchetypeAssertion`

Reply:
I personally like `Invariant`, as the mathematical definition of an invariant is quite close to the intended meaning: 

> a function, quantity, or property which remains unchanged when a specified transformation is applied.

That being said, I'll admit that most users probably won't know that definition without looking it up. My vote is for `ArchetypeAssertion`, since it lines up well with other Rust terminology.

Reply:
I also prefer `Requirement` and `Rule` over the others because a beginner will almost certainly be able to grasp what they mean. `Assertion` is still jargon, just much less so than `Invariant`.

Reply:
That's actually a very useful analysis that I hadn't considered. In that case, I'm in favor of `ArchetypeRule`, as I think `ArchetypeRequirement` could cause confusion with the `requires` method.

Reply:
My current vote is for `WorldInvariant`. I think `Rule` is not quite specific enough, and the meaning is picked up quickly in practice.

Reply:
I feel like users are likely to want to define their own `Rule` type in a lot of games. If the 'rule' terminology is the one we choose to go with, I think a better name then might be `ComponentRule`?

My preferred term for this functionality would be `Constraint`. I think it is more intuitive than `Invariant` and more accurate than `Rule`.

Reply:
Agreed on namespace conflict with `Rule`, it's too general of a name to use here.

Reply:
How do you feel about Constraint vs Invariant?

Reply:
It's a bit less correct IMO but a bit clearer. I think Constraint > Invariant > Rule, but only weakly.

Reply:
Interesting, I was thinking Constraint was more correct than Invariant. Can you explain it to me? Is it that invariant captures the idea of “anyof” etc. more accurately than constraint does?

Reply:
An invariant is a property that is constantly present despite transformations.
A constraint is just the rule itself (that in our case verifies that the invariant holds).